File: topk-py/tests/test_collection_get.py
import pytest
from topk_sdk import data, error

from . import ProjectContext
from .utils import dataset


def test_get_from_non_existent_collection(ctx: ProjectContext):
    with pytest.raises(error.CollectionNotFoundError):
        ctx.client.collection("missing").get(["doc1"])


def test_get_non_existent_document(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    docs = ctx.client.collection(collection.name).get(["missing"])
    assert docs == {}


def test_get_document(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    docs = ctx.client.collection(collection.name).get(["lotr"])

    assert docs == {
        "lotr": {
            "_id": "lotr",
            "title": "The Lord of the Rings: The Fellowship of the Ring",
            "published_year": 1954,
            "summary": "A group of unlikely heroes sets out to destroy a powerful, evil ring.",
            "summary_embedding": [9.0] * 16,
            "sparse_f32_embedding": {9: 1.0, 10: 2.0, 11: 3.0},
            "sparse_u8_embedding": {9: 1, 10: 2, 11: 3},
        }
    }


def test_get_multiple_documents(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    docs = ctx.client.collection(collection.name).get(["lotr", "moby"])

    assert docs == {
        "lotr": {
            "_id": "lotr",
            "title": "The Lord of the Rings: The Fellowship of the Ring",
            "published_year": 1954,
            "summary": "A group of unlikely heroes sets out to destroy a powerful, evil ring.",
            "summary_embedding": [9.0] * 16,
            "sparse_f32_embedding": {9: 1.0, 10: 2.0, 11: 3.0},
            "sparse_u8_embedding": {9: 1, 10: 2, 11: 3},
        },
        "moby": {
            "_id": "moby",
            "title": "Moby-Dick",
            "published_year": 1851,
            "summary": "A sailor's obsessive quest to hunt a great white whale leads to tragic consequences.",
            "summary_embedding": [6.0] * 16,
            "sparse_f32_embedding": {6: 1.0, 7: 2.0, 8: 3.0},
            "sparse_u8_embedding": {6: 1, 7: 2, 8: 3},
        },
    }


def test_get_document_fields(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    docs = ctx.client.collection(collection.name).get(
        ["lotr"],
        fields=["title", "published_year"],
    )

    assert docs == {
        "lotr": {
            "_id": "lotr",
            "title": "The Lord of the Rings: The Fellowship of the Ring",
            "published_year": 1954,
        }
    }


def test_get_empty_document_list(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    # Empty document list should raise an error
    with pytest.raises(error.InvalidArgumentError):
        ctx.client.collection(collection.name).get([])


def test_get_document_with_lsn(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    # First upsert to get an LSN - use a document that matches the schema
    lsn = ctx.client.collection(collection.name).upsert(
        [
            {
                "_id": "test_doc",
                "title": "Test Document",
                "published_year": 2023,
                "summary": "A test document for LSN testing",
                "summary_embedding": [1.0] * 16,
                "sparse_f32_embedding": {1: 1.0},
                "sparse_u8_embedding": data.u8_sparse_vector({1: 1}),
            }
        ]
    )

    # Get the document with the LSN
    docs = ctx.client.collection(collection.name).get(["test_doc"], lsn=lsn)

    assert docs == {
        "test_doc": {
            "_id": "test_doc",
            "title": "Test Document",
            "published_year": 2023,
            "summary": "A test document for LSN testing",
            "summary_embedding": [1.0] * 16,
            "sparse_f32_embedding": {1: 1.0},
            "sparse_u8_embedding": {1: 1},
        }
    }

File: topk-rs/tests/test_collection_get.rs
use std::collections::HashMap;

use test_context::test_context;
use topk_rs::data::literal;
use topk_rs::doc;
use topk_rs::proto::v1::data::ConsistencyLevel;
use topk_rs::query::field;
use topk_rs::Error;

mod utils;
use utils::dataset;
use utils::ProjectTestContext;

use crate::utils::dataset::Dataset;

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_get_from_non_existent_collection(ctx: &mut ProjectTestContext) {
    let err = ctx
        .client
        .collection("missing")
        .get(["doc1"], None, None, None)
        .await
        .expect_err("should not be able to get document from non-existent collection");

    assert!(matches!(err, Error::CollectionNotFound));
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_get_non_existent_document(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    let docs = ctx
        .client
        .collection(&collection.name)
        .get(["missing"], None, None, None)
        .await
        .expect("get failed");

    assert_eq!(docs, HashMap::new());
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_get_document(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    let docs = dataset::books::docs();
    let lotr = docs.find_by_id("lotr").unwrap().clone();

    let docs = ctx
        .client
        .collection(&collection.name)
        .get(["lotr"], None, None, None)
        .await
        .expect("could not get document");

    assert_eq!(docs, HashMap::from([("lotr".to_string(), lotr.fields)]));
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_get_multiple_documents(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    let docs = ctx
        .client
        .collection(&collection.name)
        .get(["lotr", "moby"], None, None, None)
        .await
        .expect("could not get documents");

    let books = dataset::books::docs();
    let lotr = books.find_by_id("lotr").unwrap().clone();
    let moby = books.find_by_id("moby").unwrap().clone();

    assert_eq!(
        docs,
        HashMap::from([
            ("lotr".to_string(), lotr.fields),
            ("moby".to_string(), moby.fields)
        ])
    );
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_get_document_fields(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    let docs = ctx
        .client
        .collection(&collection.name)
        .get(
            ["lotr"],
            Some(vec!["title".to_string(), "published_year".to_string()]),
            None,
            None,
        )
        .await
        .expect("could not get document");

    assert_eq!(
        docs,
        HashMap::from([(
            "lotr".to_string(),
            doc!(
                "_id" => "lotr",
                "title" => "The Lord of the Rings: The Fellowship of the Ring",
                "published_year" => 1954 as u32
            )
            .fields
        )])
    );
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_get_updated_document(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    let mut lotr = dataset::books::docs().find_by_id("lotr").unwrap().clone();

    // Update document
    lotr.fields
        .insert("published_year".to_string(), 2025.into());

    ctx.client
        .collection(&collection.name)
        .upsert(vec![lotr.clone()])
        .await
        .expect("could not upsert document");

    let docs = ctx
        .client
        .collection(&collection.name)
        .get(["lotr"], None, None, Some(ConsistencyLevel::Strong))
        .await
        .expect("could not get document");

    assert_eq!(docs, HashMap::from([("lotr".to_string(), lotr.fields)]));
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_get_deleted_document(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    ctx.client
        .collection(&collection.name)
        .delete(vec!["lotr".to_string()])
        .await
        .expect("could not upsert document");

    let docs = ctx
        .client
        .collection(&collection.name)
        .get(["lotr"], None, None, Some(ConsistencyLevel::Strong))
        .await
        .expect("could not get document");

    assert_eq!(docs, HashMap::new());
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_get_with_delete_filter(ctx: &mut ProjectTestContext) {
    let collection = ctx
        .client
        .collections()
        .create(ctx.wrap("test"), HashMap::default())
        .await
        .expect("could not create collection");

    let collection = ctx.client.collection(&collection.name);

    for batch_idx in 0..3 {
        let lsn = collection
            .upsert(
                (0..5)
                    .map(|i| {
                        let idx = batch_idx * 5 + i;
                        doc!("_id" => format!("{}", idx), "batch_idx" => batch_idx)
                    })
                    .collect(),
            )
            .await
            .expect("could not upsert document");

        assert_eq!(lsn, format!("{}", batch_idx + 1));
    }

    // Delete using filter
    let lsn = collection
        .delete(field("batch_idx").gte(literal(1)))
        .await
        .expect("could not delete document");
    assert_eq!(lsn, "4");

    // Get documents
    let docs = collection
        .get(["2", "8", "13"], None, Some(lsn.clone()), None)
        .await
        .expect("could not get documents");

    assert_eq!(docs.len(), 1, "{docs:?}");
    assert_eq!(
        docs.get("2").expect("document not found"),
        &doc!("_id" => "2", "batch_idx" => 0).fields
    );

    // Upsert more records. The upsert records satisfy the delete filter
    // but should not be deleted (snice the write happended after the delete).
    let lsn = collection
        .upsert(
            (10..15)
                .map(|i| doc!("_id" => format!("{}", i), "batch_idx" => 2, "updated" => true))
                .collect(),
        )
        .await
        .expect("could not upsert document");
    assert_eq!(lsn, "5");

    // Get documents
    let docs = collection
        .get(["2", "8", "13"], None, Some(lsn.clone()), None)
        .await
        .expect("could not get documents");

    assert_eq!(docs.len(), 2, "{docs:?}");
    assert!(docs.get("8").is_none());
    assert_eq!(
        docs.get("2").expect("document not found"),
        &doc!("_id" => "2", "batch_idx" => 0).fields
    );
    assert_eq!(
        docs.get("13").expect("document not found"),
        &doc!("_id" => "13", "batch_idx" => 2, "updated" => true).fields
    );
}
